sway/tree/layout.c

```diff
@@ -166,9 +172,183 @@ void container_move_to(struct sway_container *container,
+	// TODO: Accomodate for fullscreen view
+	int i;
+	struct sway_container *current = container;
+	struct sway_container *target = container;
+	while (1) {
+		i = index_child(current);
+		int offs_i = i + offs;
+		if (current->type == C_OUTPUT) {
+			enum wlr_direction wlr_dir;
+			sway_dir_to_wlr(dir, &wlr_dir);
+			double ref_x = current->x + current->width / 2;
+			double ref_y = current->y + current->height / 2;
+			ref_x += current->sway_output->wlr_output->lx;
+			ref_y += current->sway_output->wlr_output->ly;
+
+			struct wlr_output *next = wlr_output_layout_adjacent_output(
+				root_container.sway_root->output_layout, wlr_dir,
+				current->sway_output->wlr_output,
+				current->x, current->y);
+
+			if (!next) {
+				// Hit edge of output layout, select workspace instead
+				wlr_log(L_DEBUG, "Hit edge of output, selecting workspace.");
+				target = seat_get_focus_inactive(
+						config->handler_context.seat, target);
+				if (!target) {
+					target = current->children->items[0];
+				} else if (target->type != C_WORKSPACE) {
+					target = container_parent(target, C_WORKSPACE);
+				}
+				// If the container is already the most extreme container on
+				// this workspace, call it quits
+				int dest_i = offs > 0 ? target->children->length - 1 : 0;
+				if (target->children->items[dest_i] == container) {
+					return;
+				}
+				break;
+			}
+			struct sway_output *next_output = next->data;
+			current = target = next_output->swayc;
+			wlr_log(L_DEBUG, "Selected next output (%s)", current->name);
+			break;
+		} else if (current->type == C_WORKSPACE) {
+			if (is_perpendicular(current->layout, dir)) {
+				wlr_log(L_DEBUG, "Moving perpendicular to workspace");
+				break;
+			} else {
+				// Just skip to the output
+				current = target = current->parent;
+			}
+		} else if (!is_perpendicular(current->parent->layout, dir) &&
+				offs_i >= 0 && offs_i < current->parent->children->length) {
+			target = current->parent->children->items[offs_i];
+			wlr_log(L_DEBUG, "Selected sibling %p %s '%s'", target,
+					container_type_to_str(target->type), target->name);
+			break;
+		} else {
+			current = target = current->parent;
+			wlr_log(L_DEBUG, "Selected parent %p %s '%s'", current,
+					container_type_to_str(current->type), current->name);
+		}
+	}
+
+	wlr_log(L_DEBUG, "target: %p %s '%s'", target,
+			container_type_to_str(target->type), target->name);
+
+	struct sway_container *parent = container->parent;
+	switch (target->type) {
+	case C_VIEW:
+		if (target->parent == container->parent) {
+			wlr_log(L_DEBUG, "Swapping siblings");
+			target->parent->children->items[i + offs] = container;
+			target->parent->children->items[i] = target;
+			arrange_windows(target->parent, -1, -1);
+		} else {
+			wlr_log(L_DEBUG, "Reparenting container");
+			container_remove_child(container);
+			container_add_sibling(target, container);
+			wl_signal_emit(&container->events.reparent, parent);
```

Comment:
Maybe this signal should be triggered by the function setting `container->parent` (here `container_add_sibling`)